# Install packages

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio


In [4]:
# Import data
df = pd.read_csv('/Users/samlai/Documents/Vehicle_insurance_claim_project/data/raw/fraud_oracle.csv')

# Basic Check

In [5]:
#display all columns
pd.set_option('display.max_columns', None)
df.head()

Month  WeekOfMonth  DayOfWeek    Make AccidentArea DayOfWeekClaimed  \
0   Dec            5  Wednesday   Honda        Urban          Tuesday   
1   Jan            3  Wednesday   Honda        Urban           Monday   
2   Oct            5     Friday   Honda        Urban         Thursday   
3   Jun            2   Saturday  Toyota        Rural           Friday   
4   Jan            5     Monday   Honda        Urban          Tuesday   

  MonthClaimed  WeekOfMonthClaimed     Sex MaritalStatus  Age          Fault  \
0          Jan                   1  Female        Single   21  Policy Holder   
1          Jan                   4    Male        Single   34  Policy Holder   
2          Nov                   2    Male       Married   47  Policy Holder   
3          Jul                   1    Male       Married   65    Third Party   
4          Feb                   2  Female        Single   27    Third Party   

          PolicyType VehicleCategory     VehiclePrice  FraudFound_P  \
0  Sport - Liability           Sport  more than 69000             0   
1  Sport - Collision           Sport  more than 69000             0   
2  Sport - Collision           Sport  more than 69000             0   
3  Sedan - Liability           Sport   20000 to 29000             0   
4  Sport - Collision           Sport  more than 69000             0   

   PolicyNumber  RepNumber  Deductible  DriverRating Days_Policy_Accident  \
0             1         12         300             1         more than 30   
1             2         15         400             4         more than 30   
2             3          7         400             3         more than 30   
3             4          4         400             2         more than 30   
4             5          3         400             1         more than 30   

  Days_Policy_Claim PastNumberOfClaims AgeOfVehicle AgeOfPolicyHolder  \
0      more than 30               none      3 years          26 to 30   
1      more than 30               none      6 years          31 to 35   
2      more than 30                  1      7 years          41 to 50   
3      more than 30                  1  more than 7          51 to 65   
4      more than 30               none      5 years          31 to 35   

  PoliceReportFiled WitnessPresent AgentType NumberOfSuppliments  \
0                No             No  External                none   
1               Yes             No  External                none   
2                No             No  External                none   
3               Yes             No  External         more than 5   
4                No             No  External                none   

  AddressChange_Claim NumberOfCars  Year BasePolicy  
0              1 year       3 to 4  1994  Liability  
1           no change    1 vehicle  1994  Collision  
2           no change    1 vehicle  1994  Collision  
3           no change    1 vehicle  1994  Liability  
4           no change    1 vehicle  1994  Collision

## Missing Values check

In [6]:
df.isnull().sum().sum()

np.int64(0)

No missing values

## Check data types

In [7]:
df.head()

Month  WeekOfMonth  DayOfWeek    Make AccidentArea DayOfWeekClaimed  \
0   Dec            5  Wednesday   Honda        Urban          Tuesday   
1   Jan            3  Wednesday   Honda        Urban           Monday   
2   Oct            5     Friday   Honda        Urban         Thursday   
3   Jun            2   Saturday  Toyota        Rural           Friday   
4   Jan            5     Monday   Honda        Urban          Tuesday   

  MonthClaimed  WeekOfMonthClaimed     Sex MaritalStatus  Age          Fault  \
0          Jan                   1  Female        Single   21  Policy Holder   
1          Jan                   4    Male        Single   34  Policy Holder   
2          Nov                   2    Male       Married   47  Policy Holder   
3          Jul                   1    Male       Married   65    Third Party   
4          Feb                   2  Female        Single   27    Third Party   

          PolicyType VehicleCategory     VehiclePrice  FraudFound_P  \
0  Sport - Liability           Sport  more than 69000             0   
1  Sport - Collision           Sport  more than 69000             0   
2  Sport - Collision           Sport  more than 69000             0   
3  Sedan - Liability           Sport   20000 to 29000             0   
4  Sport - Collision           Sport  more than 69000             0   

   PolicyNumber  RepNumber  Deductible  DriverRating Days_Policy_Accident  \
0             1         12         300             1         more than 30   
1             2         15         400             4         more than 30   
2             3          7         400             3         more than 30   
3             4          4         400             2         more than 30   
4             5          3         400             1         more than 30   

  Days_Policy_Claim PastNumberOfClaims AgeOfVehicle AgeOfPolicyHolder  \
0      more than 30               none      3 years          26 to 30   
1      more than 30               none      6 years          31 to 35   
2      more than 30                  1      7 years          41 to 50   
3      more than 30                  1  more than 7          51 to 65   
4      more than 30               none      5 years          31 to 35   

  PoliceReportFiled WitnessPresent AgentType NumberOfSuppliments  \
0                No             No  External                none   
1               Yes             No  External                none   
2                No             No  External                none   
3               Yes             No  External         more than 5   
4                No             No  External                none   

  AddressChange_Claim NumberOfCars  Year BasePolicy  
0              1 year       3 to 4  1994  Liability  
1           no change    1 vehicle  1994  Collision  
2           no change    1 vehicle  1994  Collision  
3           no change    1 vehicle  1994  Liability  
4           no change    1 vehicle  1994  Collision

In [8]:
pd.set_option('display.max_columns', None)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Month                 15420 non-null  object
 1   WeekOfMonth           15420 non-null  int64 
 2   DayOfWeek             15420 non-null  object
 3   Make                  15420 non-null  object
 4   AccidentArea          15420 non-null  object
 5   DayOfWeekClaimed      15420 non-null  object
 6   MonthClaimed          15420 non-null  object
 7   WeekOfMonthClaimed    15420 non-null  int64 
 8   Sex                   15420 non-null  object
 9   MaritalStatus         15420 non-null  object
 10  Age                   15420 non-null  int64 
 11  Fault                 15420 non-null  object
 12  PolicyType            15420 non-null  object
 13  VehicleCategory       15420 non-null  object
 14  VehiclePrice          15420 non-null  object
 15  FraudFound_P          15420 non-null

# EDA

## Target variable

In [125]:
fraud_dist = df['FraudFound_P'].value_counts(normalize=True)
fig = px.bar(
    x=fraud_dist.index,
    y=fraud_dist.values,
    labels={'x': 'Fraud Found', 'y': 'Percentage'},
    text=fraud_dist.values.round(3),
    color_discrete_sequence=['#FF7D00']
)

fig.update_traces(textposition='outside')
fig.update_layout(
    title_x=0.5,
    showlegend=False
)

fig.show()

Target variable is quite imbalanced

## Create Function to plot fraud ratio by category

In [10]:
def plot_fraud_ratio_by_category(df, category_col):
    df_fraud = df[df['FraudFound_P'] == 1]
    df_fraud_count = df_fraud[category_col].value_counts()
    df_count = df[category_col].value_counts()

    #make sure we have all brands
    all_brands = df_count.index
    df_fraud_count = df_fraud_count.reindex(all_brands).fillna(0)

    df_fraud_ratio_brand = round(df_fraud_count/df_count, 2).sort_values(ascending=False)

    # 創建圖表
    fig = px.bar(
        x=df_fraud_ratio_brand.index,
        y=df_fraud_ratio_brand.values,
        title=f'Fraud Ratio by {category_col}',
        labels={'x': category_col, 'y': 'Fraud Ratio'},
        text=df_fraud_ratio_brand.values
    )

    fig.update_traces(textposition='outside')
    fig.update_layout(
        title_x=0.5,
        xaxis_tickangle=-45,
        showlegend=False
    )

    fig.show()

## *Fraud by Car Brand ('Make' column)

In [108]:
df['Make'].value_counts()

Make
Pontiac      3837
Toyota       3121
Honda        2801
Mazda        2354
Chevrolet    1681
Accura        472
Ford          450
VW            283
Dodge         109
Saab          108
Mercury        83
Saturn         58
Nisson         30
BMW            15
Jaguar          6
Porche          5
Mecedes         4
Ferrari         2
Lexus           1
Name: count, dtype: int64

In [12]:
plot_fraud_ratio_by_category(df, 'Make')


Mecedes Benz 有最高的詐欺率，比第二名的Accura 高出一倍以上，有可能是因為數量太少的關係。
Jaguar, Porche, Ferrari, Lexus的詐欺了都是0，但有可能是因為數據量太少，導致詐欺率為0。

### *考慮將數量太少的品牌合併成other

## Fraud by Vehicle Category ('VehicleCategory' column)

In [13]:
df['VehicleCategory'].value_counts()

VehicleCategory
Sedan      9671
Sport      5358
Utility     391
Name: count, dtype: int64

In [106]:
plot_fraud_ratio_by_category(df, 'VehicleCategory')

可以發現詐欺率最高的車輛類型是Utility，Sports Car 的詐欺率最低。

## *Fraud by Vihicle Price ('VehiclePrice' column)

In [15]:
df['VehiclePrice'].value_counts()

VehiclePrice
20000 to 29000     8079
30000 to 39000     3533
more than 69000    2164
less than 20000    1096
40000 to 59000      461
60000 to 69000       87
Name: count, dtype: int64

In [105]:
plot_fraud_ratio_by_category(df, 'VehiclePrice')

價格高於69,000、低於20,000及價格介於40,000-59,000的詐欺率較高。但價格位於40,000-59,000的樣本數較少，
### *可能需要考慮將40,000-59,000與60,000-69,000合併為40,000-69,000。

In [109]:
# Replace vehicle price categories
df['VehiclePrice'] = df['VehiclePrice'].replace({
    'more than 69000': 'High Price',
    '60000 to 69000': 'High Price',
    '40000 to 59000': 'Medium Price',
    '30000 to 39000': 'Medium Price', 
    '20000 to 29000': 'Low Price',
    'less than 20000': 'Low Price'
})

# Display the new distribution
print("New VehiclePrice distribution:")
print(df['VehiclePrice'].value_counts())

New VehiclePrice distribution:
VehiclePrice
Low Price       9175
Medium Price    3994
High Price      2251
Name: count, dtype: int64


In [110]:
plot_fraud_ratio_by_category(df, 'VehiclePrice')

## *Fraud by Vehicle Age ('AgeOfVehicle' column)

In [17]:
plot_fraud_ratio_by_category(df, 'AgeOfVehicle')

In [18]:
df['AgeOfVehicle'].value_counts()

AgeOfVehicle
7 years        5807
more than 7    3981
6 years        3448
5 years        1357
new             373
4 years         229
3 years         152
2 years          73
Name: count, dtype: int64

In [117]:
# Replace vehicle age categories with fewer samples
df['AgeOfVehicle'] = df['AgeOfVehicle'].replace({
    'new': '4 years or less',
    '2 years': '4 years or less', 
    '3 years': '4 years or less',
    '4 years': '4 years or less'
})

# Display the new distribution
print("New AgeOfVehicle distribution:")
print(df['AgeOfVehicle'].value_counts())


New AgeOfVehicle distribution:
AgeOfVehicle
7 years            5807
more than 7        3981
6 years            3448
5 years            1357
4 years or less     827
Name: count, dtype: int64


不看 2 years的話，較新的車越容易詐騙。但4年以下的車樣本都不多，
### *考慮將4年以下的類別合併為'4年以下'。
### *當成Oridinal Data處理

## Fraud by Accident Area ('AccidentArea' column)

In [19]:
df['AccidentArea'].value_counts()

AccidentArea
Urban    13822
Rural     1598
Name: count, dtype: int64

In [20]:
plot_fraud_ratio_by_category(df, 'AccidentArea')

Rural area詐欺率較高一些。

## *Fraud by Sex ('Sex' column) and Marital Status ('MaritalStatus' column)

In [21]:
plot_fraud_ratio_by_category(df, 'Sex')
plot_fraud_ratio_by_category(df, 'MaritalStatus')

In [22]:
df['MaritalStatus'].value_counts()

MaritalStatus
Married     10625
Single       4684
Divorced       76
Widow          35
Name: count, dtype: int64

In [23]:
df['Sex and MaritalStatus'] = df['Sex'] + ' X ' + df['MaritalStatus']
plot_fraud_ratio_by_category(df, 'Sex and MaritalStatus')

In [24]:
df['Sex and MaritalStatus'].value_counts()

Sex and MaritalStatus
Male X Married       9300
Male X Single        3653
Female X Married     1325
Female X Single      1031
Female X Divorced      40
Male X Divorced        36
Female X Widow         24
Male X Widow           11
Name: count, dtype: int64

1. 男性詐騙率較高
2. 婚姻狀態與詐騙的關係不明顯
Widow 與 Divorced的樣本數較少，
### *考慮將Widow 與 Divorced合併為其他，或是將Widow, Divorced 與 Single 合併為 Single & Other。

## Fraud by Fault ('Fault' column) --> Important Feature

In [25]:
df['Fault'].value_counts()

Fault
Policy Holder    11230
Third Party       4190
Name: count, dtype: int64

In [26]:
plot_fraud_ratio_by_category(df, 'Fault')

可以發現若過失為保單持有人，詐欺率高出非常多。若過失為第三方，只有 1% 為詐欺。

## *Fraud by Policy Type ('PolicyType' column) and Based Policy ('BasePolicy' column) --> 後者為Important Feature

In [27]:
df['PolicyType'].value_counts()

PolicyType
Sedan - Collision       5584
Sedan - Liability       4987
Sedan - All Perils      4087
Sport - Collision        348
Utility - All Perils     340
Utility - Collision       30
Sport - All Perils        22
Utility - Liability       21
Sport - Liability          1
Name: count, dtype: int64

In [28]:
plot_fraud_ratio_by_category(df, 'PolicyType')

In [29]:
df['BasePolicy'].value_counts()

BasePolicy
Collision     5962
Liability     5009
All Perils    4449
Name: count, dtype: int64

In [30]:
plot_fraud_ratio_by_category(df, 'BasePolicy')

全險的詐欺率較高，高出責任險10倍以上。
### *PolicyType Column為Based Policy與Vehicle Category的組合，應考慮該保留哪個特徵。

## Fraud by Driver Rating ('DriverRating' column)

In [31]:
df['DriverRating'].value_counts()

DriverRating
1    3944
3    3884
2    3801
4    3791
Name: count, dtype: int64

In [32]:
plot_fraud_ratio_by_category(df, 'DriverRating')

每個Driver Rating的詐欺率都差不多，這個特徵的貢獻可能不高。


## *Fraud by 'Days_Policy_Accident' and 'Days_Policy_Claim' Columns --> Both Importance Features

In [33]:
df['Days_Policy_Accident'].value_counts()

Days_Policy_Accident
more than 30    15247
none               55
8 to 15            55
15 to 30           49
1 to 7             14
Name: count, dtype: int64

In [34]:
df['Days_Policy_Claim'].value_counts()

Days_Policy_Claim
more than 30    15342
15 to 30           56
8 to 15            21
none                1
Name: count, dtype: int64

In [35]:
df.loc[df['Days_Policy_Accident'] != 'more than 30', 'Days_Policy_Accident'] = 'less than 30'
df.loc[df['Days_Policy_Claim'] != 'more than 30', 'Days_Policy_Claim'] = 'less than 30'

plot_fraud_ratio_by_category(df, 'Days_Policy_Accident')
plot_fraud_ratio_by_category(df, 'Days_Policy_Claim')

### *這兩個特徵的的各類別樣本數都很少，都將其合併為more than 30 與 less than 30。  
合併後發現兩個特徵中less than 30 的詐欺率都高出許多，隨然樣本數都不多，但可能與是否為詐欺高度相關。

## Fraud by 'PastNumberOfClaims' column --> Important Feature

In [36]:
df['PastNumberOfClaims'].value_counts()

PastNumberOfClaims
2 to 4         5485
none           4352
1              3573
more than 4    2010
Name: count, dtype: int64

In [37]:
plot_fraud_ratio_by_category(df, 'PastNumberOfClaims')

投保人之前提出索賠的次數越少，越有可能是詐欺。

## Fraud by 'PoliceReportFiled', 'WitnessPresent' and 'NumberOfSuppliments' columns --> 前兩個為 Importance Features

In [38]:
df['PoliceReportFiled'].value_counts()

PoliceReportFiled
No     14992
Yes      428
Name: count, dtype: int64

In [39]:
plot_fraud_ratio_by_category(df, 'PoliceReportFiled')

In [40]:
df['WitnessPresent'].value_counts()

WitnessPresent
No     15333
Yes       87
Name: count, dtype: int64

In [41]:
plot_fraud_ratio_by_category(df, 'WitnessPresent')

In [42]:
df['NumberOfSuppliments'].value_counts()

NumberOfSuppliments
none           7047
more than 5    3867
1 to 2         2489
3 to 5         2017
Name: count, dtype: int64

In [43]:
plot_fraud_ratio_by_category(df, 'NumberOfSuppliments')

沒有Police Report 或是沒有目擊者時，詐欺率高出許多。  
雖然有Police Report及有目擊者的樣本數都不多，但卻與詐欺高度相關，應該是重要的特徵。  
沒有補充文件時，詐欺率較高，但不明顯。


## Fraud by 'AgentType', 'AddressChange_Claim' and 'NumberOfCars' columns --> 第二個為important feature

In [44]:
df['AgentType'].value_counts()

AgentType
External    15179
Internal      241
Name: count, dtype: int64

In [45]:
plot_fraud_ratio_by_category(df, 'AgentType')

In [46]:
df['AddressChange_Claim'].value_counts()

AddressChange_Claim
no change         14324
4 to 8 years        631
2 to 3 years        291
1 year              170
under 6 months        4
Name: count, dtype: int64

In [47]:
df.loc[(df['AddressChange_Claim'] == 'under 6 months') | (df['AddressChange_Claim'] == '1 year') |
       (df['AddressChange_Claim'] == '2 to 3 years'), 'AddressChange_Claim'] = 'under 3 years'
plot_fraud_ratio_by_category(df, 'AddressChange_Claim')

In [48]:
df['NumberOfCars'].value_counts()

NumberOfCars
1 vehicle      14316
2 vehicles       709
3 to 4           372
5 to 8            21
more than 8        2
Name: count, dtype: int64

In [49]:
plot_fraud_ratio_by_category(df, 'NumberOfCars')

In [50]:
df.loc[df['NumberOfCars'] != '1 vehicle', 'NumberOfCars'] = 'more than 1 vehicle'
plot_fraud_ratio_by_category(df, 'NumberOfCars')

1. AgentType中的少數類別Internal Agent反而詐欺率較低，這個特徵的貢獻可能不高。  
### *AddressChange_Claim中部分類別樣本較少，需進行合併。  
2. 合併後發現 under 3 years的詐欺率高出許多，可能與是否為詐欺高度相關。  
3. Number of Cars與詐欺的關係不明顯。

## *Fraud by Time related features

### *Day of week & Day of Week Claim

In [51]:
df['DayOfWeek'].value_counts()

DayOfWeek
Monday       2616
Friday       2445
Tuesday      2300
Thursday     2173
Wednesday    2159
Saturday     1982
Sunday       1745
Name: count, dtype: int64

In [52]:
df['DayOfWeekClaimed'].value_counts()

DayOfWeekClaimed
Monday       3757
Tuesday      3375
Wednesday    2951
Thursday     2660
Friday       2497
Saturday      127
Sunday         52
0               1
Name: count, dtype: int64

In [53]:
plot_fraud_ratio_by_category(df, 'DayOfWeek')

In [54]:
df.loc[(df['DayOfWeekClaimed'] == 'Saturday') | (df['DayOfWeekClaimed'] == 'Sunday') |
       (df['DayOfWeekClaimed'] == '0'), 'DayOfWeekClaimed'] = 'Weekend'
plot_fraud_ratio_by_category(df, 'DayOfWeekClaimed')

In [55]:
from scipy.stats import chi2_contingency

# 創建列聯表
contingency_table = pd.crosstab(df['DayOfWeek'], df['DayOfWeekClaimed'])
# 卡方檢驗
chi2, p, dof, expected = chi2_contingency(contingency_table)
# 判斷相關性
if p < 0.05:
    print("兩個變數之間有顯著相關性")
else:
    print("兩個變數之間沒有顯著相關性")

兩個變數之間有顯著相關性


#### *兩個變數之間有顯著相關性，後續觀察是否移除其中一個變數。
#### *DayOfWeekClaimed中部分類別樣本較少，Saturday, Sunday與0合併為Weekend。  

### *Month & MonthClaimed

In [56]:
df['Month'].value_counts()

Month
Jan    1411
May    1367
Mar    1360
Jun    1321
Oct    1305
Dec    1285
Apr    1280
Feb    1266
Jul    1257
Sep    1240
Nov    1201
Aug    1127
Name: count, dtype: int64

In [57]:
df['MonthClaimed'].value_counts()

MonthClaimed
Jan    1446
May    1411
Mar    1348
Oct    1339
Jun    1293
Feb    1287
Nov    1285
Apr    1271
Sep    1242
Jul    1225
Dec    1146
Aug    1126
0         1
Name: count, dtype: int64

In [58]:
plot_fraud_ratio_by_category(df, 'Month')

In [59]:
plot_fraud_ratio_by_category(df, 'MonthClaimed')

In [60]:
from scipy.stats import chi2_contingency

# 創建列聯表
contingency_table = pd.crosstab(df['Month'], df['MonthClaimed'])
# 卡方檢驗
chi2, p, dof, expected = chi2_contingency(contingency_table)
# 判斷相關性
if p < 0.05:
    print("兩個變數之間有顯著相關性")
else:
    print("兩個變數之間沒有顯著相關性")

兩個變數之間有顯著相關性


#### *兩個變數之間有顯著相關性，後續觀察是否移除其中一個變數。

### *WeekOfMonth & WeekOfMonthClaimed

In [61]:
from scipy.stats import chi2_contingency

# 創建列聯表
contingency_table = pd.crosstab(df['WeekOfMonth'], df['WeekOfMonthClaimed'])
# 卡方檢驗
chi2, p, dof, expected = chi2_contingency(contingency_table)
# 判斷相關性
if p < 0.05:
    print("兩個變數之間有顯著相關性")
else:
    print("兩個變數之間沒有顯著相關性")

兩個變數之間有顯著相關性


#### *兩個變數之間有顯著相關性，後續觀察是否移除其中一個變數。

## *Fraud by Age & AgeOfPolicyHolder columns

In [62]:
plot_fraud_ratio_by_category(df, 'Age')

In [63]:
len(df[df['Age'] == 0])

320

In [64]:
plot_fraud_ratio_by_category(df, 'AgeOfPolicyHolder')

In [65]:
df['AgeOfPolicyHolder'].value_counts()

AgeOfPolicyHolder
31 to 35    5593
36 to 40    4043
41 to 50    2828
51 to 65    1392
26 to 30     613
over 65      508
16 to 17     320
21 to 25     108
18 to 20      15
Name: count, dtype: int64

發現25歲以下的投保人詐欺率較高。
### *考慮將25歲以下的投保人合併為'25歲以下'。但也可保留原始類別，因為可能與詐欺高度相關。
### *有320個樣本的Age是0，這不合理（因為是車禍事件的投保人）。後續要用model imputation處理。


## Fraud by 'RepNumber' & 'Deductible' columns

In [66]:
df['RepNumber'].value_counts()

RepNumber
7     1069
9      999
1      987
5      987
10     986
12     977
15     977
16     967
2      956
3      949
11     948
6      942
14     941
8      931
4      912
13     892
Name: count, dtype: int64

In [67]:
plot_fraud_ratio_by_category(df, 'RepNumber')

In [68]:
df['Deductible'].value_counts()

Deductible
400    14838
700      311
500      263
300        8
Name: count, dtype: int64

In [69]:
plot_fraud_ratio_by_category(df, 'Deductible')

各個Insurance Representative的詐欺率都差不多。  
Deductible也沒什麼規律。
### *考慮將Deductible中少數類別合併。


# 建立高風險組合表

In [118]:
# 這個dataset是在Preprocessing notebook中得到的，經過一些資料處理
df_v = pd.read_csv('../data/processed/df_for_visualisation.csv')

In [112]:
df_v.loc[df_v['Days_Policy_Claim'] != 'more than 30', 'Days_Policy_Claim'] = 'less than 30'

In [113]:
# Combine age groups with small counts
df_v['AgeOfPolicyHolder'] = df_v['AgeOfPolicyHolder'].replace({
    '16 to 17': 'under 30',
    '18 to 20': 'under 30', 
    '21 to 25': 'under 30',
    '26 to 30': 'under 30',
    '31 to 35': '31 to 35',
    '36 to 40': '36 to 40',
    '41 to 50': '41 to 50',
    '51 to 65': '51 and above',
    'over 65': '51 and above'
})

df_v['AgeOfPolicyHolder'].value_counts()

AgeOfPolicyHolder
31 to 35        5593
36 to 40        4043
41 to 50        2828
51 and above    1900
under 30        1056
Name: count, dtype: int64

In [114]:
# Replace vehicle price categories
df_v['VehiclePrice'] = df_v['VehiclePrice'].replace({
    'more than 69000': 'High Price',
    '60000 to 69000': 'High Price',
    '40000 to 59000': 'Medium Price',
    '30000 to 39000': 'Medium Price', 
    '20000 to 29000': 'Low Price',
    'less than 20000': 'Low Price'
})

In [119]:
# Replace vehicle age categories with fewer samples
df_v['AgeOfVehicle'] = df_v['AgeOfVehicle'].replace({
    'new': '4 years or less',
    '2 years': '4 years or less', 
    '3 years': '4 years or less',
    '4 years': '4 years or less'
})


In [120]:
# export new df_for_visualisation
df_v.to_csv('../data/processed/df_for_visualisation.csv', index = False)
df_v.to_csv('../app/data/df_for_visualisation.csv', index = False)

In [124]:
# Create aggregation table
risk_groups = df_v.groupby(['AgeOfPolicyHolder', 'BasePolicy', 'VehiclePrice']).agg(
    total_claims=('FraudFound_P', 'count'),
    fraud_claims=('FraudFound_P', 'sum')
).reset_index()

# Calculate fraud rate
risk_groups['Fraud_Rate'] = risk_groups['fraud_claims'] / risk_groups['total_claims']

# Sort by fraud rate in descending order
risk_groups = risk_groups.sort_values('Fraud_Rate', ascending=False)

risk_groups.head(30)

AgeOfPolicyHolder  BasePolicy  VehiclePrice  total_claims  fraud_claims  \
36          under 30  All Perils    High Price           122            25   
0           31 to 35  All Perils    High Price           163            27   
39          under 30   Collision    High Price           156            23   
1           31 to 35  All Perils     Low Price           961           121   
21          41 to 50   Collision    High Price           133            15   
9           36 to 40  All Perils    High Price           212            23   
2           31 to 35  All Perils  Medium Price           219            23   
10          36 to 40  All Perils     Low Price           689            72   
27      51 and above  All Perils    High Price           147            15   
37          under 30  All Perils     Low Price           163            16   
12          36 to 40   Collision    High Price           185            18   
3           31 to 35   Collision    High Price           236            22   
14          36 to 40   Collision  Medium Price           464            42   
11          36 to 40  All Perils  Medium Price           211            19   
5           31 to 35   Collision  Medium Price           561            47   
18          41 to 50  All Perils    High Price           188            15   
29      51 and above  All Perils  Medium Price           139            11   
28      51 and above  All Perils     Low Price           465            35   
23          41 to 50   Collision  Medium Price           297            22   
30      51 and above   Collision    High Price            82             6   
32      51 and above   Collision  Medium Price           170            12   
19          41 to 50  All Perils     Low Price           556            38   
4           31 to 35   Collision     Low Price          1627           111   
22          41 to 50   Collision     Low Price           535            34   
20          41 to 50  All Perils  Medium Price           197            12   
13          36 to 40   Collision     Low Price           848            50   
31      51 and above   Collision     Low Price           354            19   
41          under 30   Collision  Medium Price            57             3   
44          under 30   Liability  Medium Price            23             1   
40          under 30   Collision     Low Price           257            11   

    Fraud_Rate  
36    0.204918  
0     0.165644  
39    0.147436  
1     0.125911  
21    0.112782  
9     0.108491  
2     0.105023  
10    0.104499  
27    0.102041  
37    0.098160  
12    0.097297  
3     0.093220  
14    0.090517  
11    0.090047  
5     0.083779  
18    0.079787  
29    0.079137  
28    0.075269  
23    0.074074  
30    0.073171  
32    0.070588  
19    0.068345  
4     0.068224  
22    0.063551  
20    0.060914  
13    0.058962  
31    0.053672  
41    0.052632  
44    0.043478  
40    0.042802